platform/darwin/src/MGLStyle.mm

```diff
@@ -282,18 +282,19 @@ - (void)insertObject:(MGLStyleLayer *)styleLayer inLayersAtIndex:(NSUInteger)ind
          styleLayer];
     }
     auto layers = self.mapView.mbglMap->getLayers();
-    if (index > layers.size()) {
+    if (index >= layers.size()) {
         [NSException raise:NSRangeException
                     format:@"Cannot insert style layer at out-of-bounds index %lu.", (unsigned long)index];
     } else if (index == 0) {
         try {
-            [styleLayer addToMapView:self.mapView belowLayer:nil];
+            MGLStyleLayer *sibling = layers.size() ? [self layerFromMBGLLayer:layers.at(0)] : nil;
+            [styleLayer addToMapView:self.mapView belowLayer:sibling];
         } catch (const std::runtime_error & err) {
             [NSException raise:@"MGLRedundantLayerIdentifierException" format:@"%s", err.what()];
         }
     } else {
         try {
-            MGLStyleLayer *sibling = [self layerFromMBGLLayer:layers.at(layers.size() - index)];
+            MGLStyleLayer *sibling = [self layerFromMBGLLayer:layers.at(index + 1)];
```

Comment:
Inserting a layer at an index should put it below (before) the layer currently at that index. Therefore, this should be `layers.at(index)`.